managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -641,6 +641,7 @@ public void recoverIndividualDeletedMessages(PositionInfo positionInfo) {
         if (positionInfo.getIndividualDeletedMessagesCount() > 0) {
             recoverIndividualDeletedMessages(positionInfo.getIndividualDeletedMessagesList());
         } else if (positionInfo.getIndividualDeletedMessageRangesCount() > 0) {
+            lock.writeLock().lock();
```

Comment:
I guess this should be just before the `try {` block so that there isn't a risk of the lock remaining locked when an exception happens.